File: backend/app/schemas/portal.py
from __future__ import annotations

import uuid
from datetime import date, datetime, time
from typing import Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator, model_validator

class CasePortalFolderAccessGrantOut(BaseModel):
    folder_path: str
    contact_id: uuid.UUID
    contact_name: str

class CasePortalFolderShareContactOut(BaseModel):
    case_contact_id: uuid.UUID
    contact_id: uuid.UUID
    contact_name: str
    has_grant: bool
    grant_id: uuid.UUID | None
    portal_access_active: bool = True
    # Matter snapshot e-mail, then global contact card (empty when missing).
    email: str = ""
    matter_contact_type: str = ""
    is_exchange_contact: bool = False

class PortalConfigOut(BaseModel):
    firm_name: str
    portal_title: str
    portal_logo_url: str | None = None
    portal_background_color: str | None = None
    powered_by_label: str = "Powered by Canary Legal Software"
    powered_by_url: str = "https://canarylegalsoftware.co.uk"

class PortalAuthIn(BaseModel):
    access_code: str = Field(min_length=8, max_length=64)

class PortalGrantSummaryOut(BaseModel):
    id: uuid.UUID
    case_id: uuid.UUID
    case_title: str
    folder_path: str
    folder_label: str
    label: str
    can_download: bool
    can_upload: bool
    new_file_count: int = 0
    last_viewed_at: datetime | None = None

class PortalAuthOut(BaseModel):
    session_token: str
    contact_name: str
    grants: list[PortalGrantSummaryOut]
    focus_case_id: uuid.UUID | None = None
    staff_preview: bool = False
    audience: Literal["client", "exchange"] = "client"

class PortalSessionOut(BaseModel):
    contact_name: str
    grants: list[PortalGrantSummaryOut]
    staff_preview: bool = False
    audience: Literal["client", "exchange"] = "client"
    focus_case_id: uuid.UUID | None = None

class PortalFileOut(BaseModel):
    id: uuid.UUID
    original_filename: str
    mime_type: str
    size_bytes: int
    folder_path: str
    folder_display: str = ""
    created_at: datetime
    updated_at: datetime
    is_new: bool = False

class PortalBrowseOut(BaseModel):
    subfolder: str
    breadcrumb: list[str]
    subfolders: list[str]
    files: list[PortalFileOut]
    pending_approvals: list[PortalQuoteDeliveryViewOut] = []
    pending_docusign_signings: list["PortalDocusignSigningOut"] = []
    pending_canary_signings: list["PortalCanarySignOut"] = []
    pending_portal_forms: list["PortalFormPendingOut"] = []
    new_file_count: int = 0
    last_viewed_at: datetime | None = None

class PortalFormTemplateFieldIn(BaseModel):
    field_key: str = Field(min_length=1, max_length=80)
    label: str = Field(min_length=1, max_length=500)
    field_type: Literal["section", "text", "textarea", "date", "select", "file"]
    help_text: str | None = Field(default=None, max_length=2000)
    required: bool = False
    sort_order: int = 0
    select_options: list[str] = Field(default_factory=list)

class PortalFormTemplateFieldOut(PortalFormTemplateFieldIn):
    id: uuid.UUID

class PortalFormTemplateOut(BaseModel):
    id: uuid.UUID
    name: str
    reference: str
    description: str | None = None
    matter_head_type_id: uuid.UUID | None = None
    matter_sub_type_id: uuid.UUID | None = None
    scope_summary: str = ""
    field_count: int = 0
    created_at: datetime
    updated_at: datetime

class PortalFormTemplateDetailOut(PortalFormTemplateOut):
    fields: list[PortalFormTemplateFieldOut] = []

class PortalFormTemplateCreate(BaseModel):
    name: str = Field(min_length=1, max_length=300)
    reference: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=5000)
    matter_head_type_id: uuid.UUID | None = None
    matter_sub_type_id: uuid.UUID | None = None
    fields: list[PortalFormTemplateFieldIn] = Field(default_factory=list)

class PortalFormTemplateUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=300)
    reference: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=5000)
    matter_head_type_id: uuid.UUID | None = None
    matter_sub_type_id: uuid.UUID | None = None
    fields: list[PortalFormTemplateFieldIn] | None = None

class PortalFormSendIn(BaseModel):
    template_id: uuid.UUID
    contact_id: uuid.UUID

class PortalFormSubmissionOut(BaseModel):
    id: uuid.UUID
    case_id: uuid.UUID
    template_id: uuid.UUID
    template_name: str
    template_reference: str
    contact_id: uuid.UUID
    contact_name: str
    status: str
    responses: dict = {}
    snapshot_file_id: uuid.UUID | None = None
    snapshot_filename: str = ""
    sent_at: datetime
    completed_at: datetime | None = None
    voided_at: datetime | None = None
    email_sent: bool = False
    email_skip_reason: str | None = None

class PortalFormFieldOut(BaseModel):
    field_key: str
    label: str
    field_type: str
    help_text: str | None = None
    required: bool
    sort_order: int
    select_options: list[str] = []

class PortalFormPendingOut(BaseModel):
    id: uuid.UUID
    template_name: str
    template_reference: str
    status: str
    sent_at: datetime
    case_id: uuid.UUID | None = None
    matter_label: str = ""

class PortalFormDetailOut(PortalFormSubmissionOut):
    description: str | None = None
    fields: list[PortalFormFieldOut] = []

class PortalFormSubmitIn(BaseModel):
    responses: dict = Field(default_factory=dict)

class PortalOtpRequestIn(BaseModel):
    email: str = Field(min_length=3, max_length=320)

class PortalOtpVerifyIn(BaseModel):
    email: str = Field(min_length=3, max_length=320)
    code: str = Field(min_length=4, max_length=12)

class CasePortalNotifyFilesIn(BaseModel):
    folder_path: str = ""
    filenames: list[str] = Field(min_length=1)

class CasePortalNotifyFilesOut(BaseModel):
    contacts_notified: int
    alerts_skipped_reason: str | None = None

class CasePortalActivityOut(BaseModel):
    id: uuid.UUID
    action: str
    summary: str
    contact_name: str | None
    created_at: datetime

class CasePortalShareStatusOut(BaseModel):
    portal_enabled: bool
    active_grant_count: int
    contact_count: int

class CasePortalStaffUserOut(BaseModel):
    id: uuid.UUID
    display_name: str
    email: str

class CasePortalNotificationSettingsOut(BaseModel):
    staff_user_ids: list[uuid.UUID]
    staff_users: list[CasePortalStaffUserOut] = Field(default_factory=list)

class CasePortalNotificationSettingsIn(BaseModel):
    staff_user_ids: list[uuid.UUID] = Field(default_factory=list)

class CasePortalPreviewContactOut(BaseModel):
    contact_id: uuid.UUID
    contact_name: str
    shared_folder_count: int
    pending_quote_count: int = 0
    pending_form_count: int = 0
    pending_canary_sign_count: int = 0

class CasePortalPreviewIn(BaseModel):
    contact_id: uuid.UUID

class CasePortalPreviewOut(BaseModel):
    exchange_token: str
    contact_name: str
    preview_url: str

class PortalPreviewExchangeIn(BaseModel):
    exchange_token: str = Field(min_length=10, max_length=4096)

class PortalQuoteExchangeIn(BaseModel):
    exchange_token: str = Field(min_length=10, max_length=4096)

class PortalFormExchangeIn(BaseModel):
    exchange_token: str = Field(min_length=10, max_length=4096)

class PortalQuoteDeliveryViewOut(BaseModel):
    id: uuid.UUID
    file_id: uuid.UUID
    grant_id: uuid.UUID | None
    case_id: uuid.UUID | None = None
    case_title: str = ""
    original_filename: str
    mime_type: str = "application/octet-stream"
    size_bytes: int = 0
    folder_display: str = ""
    status: str
    can_respond: bool
    decline_reason: str | None = None
    responded_at: datetime | None = None
    portal_pdf_available: bool = False

class PortalQuoteRespondIn(BaseModel):
    accepted: bool
    decline_reason: str | None = Field(default=None, max_length=2000)

class PortalQuoteExchangeOut(BaseModel):
    session_token: str
    contact_name: str
    grants: list[PortalGrantSummaryOut]
    quote: PortalQuoteDeliveryViewOut

class PortalFormExchangeOut(BaseModel):
    session_token: str
    contact_name: str
    grants: list[PortalGrantSummaryOut]
    form: PortalFormPendingOut

class QuotePortalSendPreflightOut(BaseModel):
    alerts_configured: bool

class SendQuoteViaPortalIn(BaseModel):
    contact_id: uuid.UUID

class PortalQuoteTagUpdate(BaseModel):
    is_portal_quote: bool

class QuotePortalDeliveryOut(BaseModel):
    id: uuid.UUID
    file_id: uuid.UUID
    contact_id: uuid.UUID
    contact_name: str
    status: str
    sent_at: datetime
    responded_at: datetime | None = None
    decline_reason: str | None = None
    file_version_at_send: int
    email_sent: bool = False
    email_skip_reason: str | None = None
    portal_pdf_generated: bool = False

class PublishComposeIn(BaseModel):
    notify_portal_contacts: bool = False

class ContactPortalNotificationPrefsOut(BaseModel):
    notify_files_added: bool
    notify_folder_shared: bool

class ContactPortalNotificationPrefsIn(BaseModel):
    notify_files_added: bool | None = None
    notify_folder_shared: bool | None = None

class ContactPortalAccessOut(BaseModel):
    enabled: bool
    expires_at: datetime | None
    last_login_at: datetime | None
    locked_until: datetime | None
    has_access: bool
    access_code: str | None = None
    access_record_exists: bool = False
    notify_files_added: bool = True
    notify_folder_shared: bool = True

class ContactPortalAccessCreateOut(BaseModel):
    access_code: str
    enabled: bool
    expires_at: datetime | None
    email_sent: bool = False
    email_skip_reason: str | None = None

class ContactPortalAccessActionIn(BaseModel):
    send_email: bool = False
    """When set (matter contact UI), portal must be enabled on that matter before granting access."""
    case_id: uuid.UUID | None = None

class MatterPortalAccessOut(BaseModel):
    enabled: bool
    expires_at: datetime | None
    last_login_at: datetime | None
    locked_until: datetime | None
    has_access: bool
    access_code: str | None = None
    access_record_exists: bool = False
    notify_folder_shared: bool = True
    case_id: uuid.UUID
    contact_id: uuid.UUID

class MatterPortalAccessCreateOut(BaseModel):
    access_code: str
    enabled: bool
    expires_at: datetime | None = None
    email_sent: bool = False
    email_skip_reason: str | None = None
    case_id: uuid.UUID
    contact_id: uuid.UUID

class MatterPortalAccessActionIn(BaseModel):
    send_email: bool = False

class ContactPortalAccessEmailIn(BaseModel):
    access_code: str = Field(min_length=4, max_length=64)

class ContactPortalAccessUpdateIn(BaseModel):
    enabled: bool | None = None
    expires_at: datetime | None = None

class ContactPortalGrantOut(BaseModel):
    id: uuid.UUID
    contact_id: uuid.UUID
    case_id: uuid.UUID
    case_title: str
    folder_path: str
    label: str | None
    can_download: bool
    can_upload: bool
    expires_at: datetime | None
    created_at: datetime
    email_sent: bool = False
    email_skip_reason: str | None = None

class ContactPortalGrantCreateIn(BaseModel):
    case_id: uuid.UUID
    folder_path: str = ""
    label: str | None = Field(default=None, max_length=300)
    can_download: bool = True
    can_upload: bool = True
    expires_at: datetime | None = None
    send_email: bool = False

class ContactPortalGrantUpdateIn(BaseModel):
    folder_path: str | None = None
    label: str | None = Field(default=None, max_length=300)
    can_download: bool | None = None
    can_upload: bool | None = None
    expires_at: datetime | None = None

File: frontend/src/types/portal.ts
import type { PortalCanarySignOut } from './canary_sign'
import type { PortalDocusignSigningOut } from './docusign'

export type PortalFormSubmissionSummary = {
  id: string
  status: 'pending' | 'completed' | 'voided' | 'superseded'
  contact_name: string
  template_name: string
  sent_at: string
  completed_at: string | null
}

export type QuotePortalDeliverySummary = {
  id: string
  status: 'pending' | 'accepted' | 'declined' | 'superseded'
  contact_name: string
  sent_at: string
  responded_at: string | null
  decline_reason: string | null
}

export type QuotePortalDeliveryOut = {
  id: string
  file_id: string
  contact_id: string
  contact_name: string
  status: string
  sent_at: string
  responded_at: string | null
  decline_reason: string | null
  file_version_at_send: number
  email_sent: boolean
  email_skip_reason: string | null
  portal_pdf_generated?: boolean
}

export type QuotePortalSendPreflightOut = {
  alerts_configured: boolean
}

export type PortalQuoteDeliveryViewOut = {
  id: string
  file_id: string
  grant_id: string | null
  case_id?: string | null
  case_title?: string
  original_filename: string
  mime_type?: string
  size_bytes?: number
  folder_display?: string
  status: string
  can_respond: boolean
  decline_reason: string | null
  responded_at: string | null
  portal_pdf_available?: boolean
}

export type PortalQuoteExchangeOut = {
  session_token: string
  contact_name: string
  grants: PortalGrantSummaryOut[]
  quote: PortalQuoteDeliveryViewOut
}

export type PortalFormExchangeOut = {
  session_token: string
  contact_name: string
  grants: PortalGrantSummaryOut[]
  form: PortalFormPendingOut
}

export type PortalGrantSummaryOut = {
  id: string
  case_id: string
  case_title: string
  folder_path: string
  folder_label: string
  label: string
  can_download: boolean
  can_upload: boolean
  new_file_count?: number
  last_viewed_at?: string | null
}

export type PortalAuthOut = {
  session_token: string
  contact_name: string
  grants: PortalGrantSummaryOut[]
  focus_case_id?: string | null
  staff_preview?: boolean
  audience?: 'client' | 'exchange'
}

export type PortalSessionOut = {
  contact_name: string
  grants: PortalGrantSummaryOut[]
  staff_preview?: boolean
  audience?: 'client' | 'exchange'
  focus_case_id?: string | null
}

export type PortalFileOut = {
  id: string
  original_filename: string
  mime_type: string
  size_bytes: number
  folder_path: string
  folder_display?: string
  created_at: string
  updated_at: string
  is_new?: boolean
}

export type PortalBrowseOut = {
  subfolder: string
  breadcrumb: string[]
  subfolders: string[]
  files: PortalFileOut[]
  pending_approvals?: PortalQuoteDeliveryViewOut[]
  pending_docusign_signings?: PortalDocusignSigningOut[]
  pending_canary_signings?: PortalCanarySignOut[]
  pending_portal_forms?: PortalFormPendingOut[]
  new_file_count?: number
  last_viewed_at?: string | null
}

export type PortalFormFieldType = 'section' | 'text' | 'textarea' | 'date' | 'select' | 'file'

export type PortalFormTemplateFieldIn = {
  field_key: string
  label: string
  field_type: PortalFormFieldType
  help_text?: string | null
  required?: boolean
  sort_order?: number
  select_options?: string[]
}

export type PortalFormTemplateFieldOut = PortalFormTemplateFieldIn & {
  id: string
}

export type PortalFormTemplateOut = {
  id: string
  name: string
  reference: string
  description?: string | null
  matter_head_type_id?: string | null
  matter_sub_type_id?: string | null
  scope_summary: string
  field_count: number
  created_at: string
  updated_at: string
}

export type PortalFormTemplateDetailOut = PortalFormTemplateOut & {
  fields: PortalFormTemplateFieldOut[]
}

export type PortalFormSubmissionOut = {
  id: string
  case_id: string
  template_id: string
  template_name: string
  template_reference: string
  contact_id: string
  contact_name: string
  status: string
  responses: Record<string, unknown>
  snapshot_file_id?: string | null
  snapshot_filename: string
  sent_at: string
  completed_at?: string | null
  voided_at?: string | null
  email_sent?: boolean
  email_skip_reason?: string | null
}

export type PortalFormFieldOut = {
  field_key: string
  label: string
  field_type: PortalFormFieldType
  help_text?: string | null
  required: boolean
  sort_order: number
  select_options?: string[]
}

export type PortalFormPendingOut = {
  id: string
  template_name: string
  template_reference: string
  status: string
  sent_at: string
  case_id?: string | null
  matter_label: string
}

export type PortalFormDetailOut = PortalFormSubmissionOut & {
  description?: string | null
  fields: PortalFormFieldOut[]
}

export type CasePortalActivityOut = {
  id: string
  action: string
  summary: string
  contact_name: string | null
  created_at: string
}

export type CasePortalStaffUserOut = {
  id: string
  display_name: string
  email: string
}

export type CasePortalNotificationSettingsOut = {
  staff_user_ids: string[]
  staff_users?: CasePortalStaffUserOut[]
}

export type CasePortalShareStatusOut = {
  portal_enabled: boolean
  active_grant_count: number
  contact_count: number
}

export type CasePortalPreviewContactOut = {
  contact_id: string
  contact_name: string
  shared_folder_count: number
  pending_quote_count?: number
  pending_form_count?: number
  pending_canary_sign_count?: number
}

export type CasePortalPreviewOut = {
  exchange_token: string
  contact_name: string
  preview_url: string
}

export type CasePortalNotifyFilesOut = {
  contacts_notified: number
  alerts_skipped_reason?: string | null
}

export type ContactPortalNotificationPrefsOut = {
  notify_files_added: boolean
  notify_folder_shared: boolean
}

export const PORTAL_ALERTS_NOT_CONFIGURED_MSG =
  'Automated e-mail is not configured. Ask an administrator to enable Admin → E-mail → "Enable automated alert e-mail" and set up Graph or SMTP.'

export type ContactPortalAccessOut = {
  enabled: boolean
  expires_at: string | null
  last_login_at: string | null
  locked_until: string | null
  has_access: boolean
  access_code?: string | null
  access_record_exists?: boolean
  notify_files_added?: boolean
  notify_folder_shared?: boolean
}

export type ContactPortalAccessCreateOut = {
  access_code: string
  enabled: boolean
  expires_at: string | null
  email_sent?: boolean
  email_skip_reason?: string | null
}

export type ContactPortalGrantOut = {
  id: string
  contact_id: string
  case_id: string
  case_title: string
  folder_path: string
  label: string | null
  can_download: boolean
  can_upload: boolean
  expires_at: string | null
  created_at: string
  email_sent?: boolean
  email_skip_reason?: string | null
}

export type ContactPortalGrantCreateIn = {
  case_id: string
  folder_path?: string
  label?: string | null
  can_download?: boolean
  can_upload?: boolean
  expires_at?: string | null
  send_email?: boolean
}

export type CasePortalFolderShareContactOut = {
  case_contact_id: string
  contact_id: string
  contact_name: string
  has_grant: boolean
  grant_id: string | null
  /** False when the contact will be auto-granted portal access on send. */
  portal_access_active?: boolean
  /** Matter snapshot or global contact e-mail (empty when missing). */
  email?: string
  matter_contact_type?: string
  is_exchange_contact?: boolean
}

export type MatterPortalAccessOut = {
  enabled: boolean
  expires_at: string | null
  last_login_at: string | null
  locked_until: string | null
  has_access: boolean
  access_code: string | null
  access_record_exists: boolean
  notify_folder_shared: boolean
  case_id: string
  contact_id: string
}

export type MatterPortalAccessCreateOut = {
  access_code: string
  enabled: boolean
  expires_at?: string | null
  email_sent: boolean
  email_skip_reason?: string | null
  case_id: string
  contact_id: string
}

export type CasePortalFolderAccessGrantOut = {
  folder_path: string
  contact_id: string
  contact_name: string
}
